utility/session_test.go

```diff
@@ -0,0 +1,531 @@
+	require.Error(t, err)
+}
+
+func TestSession_GetSession_InvalidFutureSession(t *testing.T) {
+	runtimeCfg, utilityMod, persistenceMod := prepareEnvironment(t, 5, 1, 1, 1)
+
+	// Test parameters
+	relayChain := test_artifacts.DefaultChains[0]
+	geoZone := "unused_geo"
+	app := runtimeCfg.GetGenesis().Applications[0]
+
+	// Local variable to keep track of the height we're getting a session for
+	currentHeight := int64(0)
+
+	// Successfully get a session for 1 block ahead of the latest committed height
+	session, err := utilityMod.GetSession(app.Address, currentHeight+1, relayChain, geoZone)
+	require.NoError(t, err)
+	require.Equal(t, currentHeight+1, session.SessionHeight)
+
+	// Expect an error for a few heights into the future
+	for height := currentHeight + 2; height < 10; height++ {
+		_, err := utilityMod.GetSession(app.Address, height, relayChain, geoZone)
+		require.Error(t, err)
+	}
+
+	// Commit new blocks for all the heights that failed above
+	for ; currentHeight < 10; currentHeight++ {
+		writeCtx, err := persistenceMod.NewRWContext(currentHeight + 1)
+		require.NoError(t, err)
+		err = writeCtx.Commit([]byte(fmt.Sprintf("proposer_height_%d", currentHeight)), []byte(fmt.Sprintf("quorum_cert_height_%d", currentHeight)))
+		require.NoError(t, err)
+		writeCtx.Release()
+	}
+
+	// Expect no errors since those blocks exist now
+	// Note that we can get the session for latest_committed + 1
+	for height := int64(1); height <= currentHeight+1; height++ {
+		_, err := utilityMod.GetSession(app.Address, height, relayChain, geoZone)
+		require.NoError(t, err)
+	}
+
+	// Verify that currentHeight + 2 fails
+	_, err = utilityMod.GetSession(app.Address, currentHeight+2, relayChain, geoZone)
+	require.Error(t, err)
+}
+
+func TestSession_GetSession_ServicersAndFishermenCounts_TotalAvailability(t *testing.T) {
+	// Prepare an environment with a lot of servicers and fishermen
+	numServicers := 100
+	numFishermen := 100
+	runtimeCfg, utilityMod, persistenceMod := prepareEnvironment(t, 5, numServicers, 1, numFishermen)
+
+	// Vary the number of actors per session using gov params and check that the session is populated with the correct number of actorss
+	tests := []struct {
+		name                   string
+		numServicersPerSession int64
+		numFishermanPerSession int64
+		wantServicerCount      int
+		wantFishermanCount     int
+	}{
+		{
+			name:                   "more actors per session than available in network",
+			numServicersPerSession: int64(numServicers) * 10,
+			numFishermanPerSession: int64(numFishermen) * 10,
+			wantServicerCount:      numServicers,
+			wantFishermanCount:     numFishermen,
+		},
+		{
+			name:                   "less actors per session than available in network",
+			numServicersPerSession: int64(numServicers) / 2,
+			numFishermanPerSession: int64(numFishermen) / 2,
+			wantServicerCount:      numServicers / 2,
+			wantFishermanCount:     numFishermen / 2,
```

Comment:
Am I'm missing something subtle or is there a copy/pasta here? Isn't this equivalent to "same number of actors..."?

Reply:
I renamed the variables, is it self explanatory now?

```go
			numServicersPerSession: int64(numStakedServicers) / 2,
			numFishermanPerSession: int64(numStakedFishermen) / 2,
			wantServicerCount:      numStakedServicers / 2,
			wantFishermanCount:     numStakedFishermen / 2,
```

If not, I can add a comment providing more details.